File: chapter 07/recipe 7.7/DniNetStructsVB/DniNetClassesObj.vb
Imports System.Runtime.InteropServices

'defines the public API for an account object
Public Interface IAccount
    Property AccountId() As Integer

    Property AccountName() As String

    Function GetBalance() As _
        <MarshalAs(UnmanagedType.Currency)> Decimal

    <ComVisible(False)> _
    Sub SetBalance(<MarshalAs(UnmanagedType.Currency)> _
        ByVal value As Decimal)
End Interface

'defines an account class
<ComVisible(False)> _
Public Class Account
    Implements IAccount

    Private m_AccountId As Integer
    Private m_AccountName As String
    Private m_Balance As Decimal

    Public Property AccountId() As Integer _
        Implements IAccount.AccountId
        Get
            Return m_AccountId
        End Get
        Set(ByVal value As Integer)
            m_AccountId = value
        End Set
    End Property

    Public Property AccountName() As String _
        Implements IAccount.AccountName
        Get
            Return m_AccountName
        End Get
        Set(ByVal value As String)
            m_AccountName = value
        End Set
    End Property

    Public Function GetBalance() As Decimal _
        Implements IAccount.GetBalance
        Return m_Balance
    End Function

    Public Sub SetBalance(ByVal value As Decimal) _
        Implements IAccount.SetBalance
        m_Balance = value
    End Sub

End Class

'define an interface for account services
Public Interface IAccountLookup
    Function RetrieveAccount(ByVal acctId As Integer) _
        As IAccount
    Sub UpdateBalance(ByRef account As IAccount)
End Interface

<ClassInterface(ClassInterfaceType.None)> _
Public Class DniNetClassesObj
    Implements IAccountLookup

    Public Function RetrieveAccount(ByVal acctId As Integer) _
        As IAccount Implements IAccountLookup.RetrieveAccount
        Dim result As Account = Nothing
        If acctId = 123 Then
            result = New Account()
            result.AccountId = acctId
            result.AccountName = "myAccount"
            result.SetBalance(1009.95D)
        End If
        Return result
    End Function

    Public Sub UpdateBalance(ByRef account As IAccount) _
        Implements IAccountLookup.UpdateBalance
        If Not account Is Nothing Then
            account.SetBalance(account.GetBalance() + 500D)
        End If
    End Sub
End Class

File: chapter 07/recipe 7.7/DniNetStructs/DniNetClassesObj.cs
using System;
using System.Runtime.InteropServices;

namespace DniNetStructs
{
	//Implement an account class
	[ComVisible(false)]
	public class Account : IAccount
	{
		private int			m_AccountId;
		private string		m_AccountName;
		private decimal		m_Balance;

		public int AccountId
		{
			get { return m_AccountId;}
			set { m_AccountId = value; }
		}

		public string AccountName
		{
			get { return m_AccountName; }
			set { m_AccountName = value; }
		}

		public decimal Balance
		{
			get { return m_Balance; }
			set { m_Balance = value; }
		}
	}

	//define an interface to expose class properties
	//to COM clients
	public interface IAccount
	{
		int AccountId		{get;}
		string AccountName	{get;}
		decimal Balance 
		{
			//marshal the balance to com as CURRENCY
			[return: MarshalAs(UnmanagedType.Currency)] get;
			//prohibit setting of balance property via com
			[ComVisible(false)]
			set;
		}
	}

	//define an interface for account services
	public interface IAccountLookup
	{
		IAccount RetrieveAccount(int acctId);
		void UpdateBalance(IAccount account);
	}

	//implement a class to provide account services
	[ClassInterface(ClassInterfaceType.None)]
	public class DniNetClassesObj : IAccountLookup
	{
		public IAccount RetrieveAccount(int acctId)
		{
			Account result = null;
			if (acctId == 123)
			{
				result = new Account();
				result.AccountId = acctId;
				result.AccountName = "myAccount";
				result.Balance = 1009.95M;
			}
			return result;
		}

		public void UpdateBalance(IAccount account)
		{
			//update the balance
			if (account != null)
			{
				account.Balance += 500.00M;
			}
		}
	}
}
